Partition sessionStorage by top-level origin

sysrqb opened the pull request:
#### d5739b8e09742e6df2e40e0691a98f1801280db3
<pre>
Partition sessionStorage by top-level origin
<a href="https://bugs.webkit.org/show_bug.cgi?id=210776">https://bugs.webkit.org/show_bug.cgi?id=210776</a>
rdar://57674840

Reviewed by Alex Christensen and Sihui Liu.

A page could access its origin&apos;s sessionStorage in both first-party and
third-party contexts. This differs from other storage systems and caches where
in a third-party context those storage areas are partitioned based on the
first-party (top-level) origin. This issue is addressed by ensuring that a
Window&apos;s sessionStorage takes the top-level origin into account at access-time.

This accounting is accomplished by keying the StorageNamespace by top-level
origin and the namespace identifier. One complexity was how the Page object
caches the resulting StorageNamespace, now partitioned by origin. This change
interacted poorly with the creation and destruction of the Page object during
navigation. As a result, ownership of the StorageNamespace is moved from the
Page into the StorageNamespaceProvider.

* LayoutTests/http/tests/security/cross-origin-session-storage-third-party-partitioned-expected.txt: Added.
* LayoutTests/http/tests/security/cross-origin-session-storage-third-party-partitioned.html: Added.
* LayoutTests/http/tests/security/resources/cross-origin-iframe-for-partitioned-session-storage.html: Added.
* LayoutTests/http/tests/security/resources/cross-origin-main-frame-for-partitioned-session-storage.html: Added.
* Source/WebCore/inspector/agents/InspectorDOMStorageAgent.cpp:
(WebCore::InspectorDOMStorageAgent::findStorageArea):
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyStorageNamespaceProvider::sessionStorageNamespace):
(WebCore::EmptyStorageNamespaceProvider::createSessionStorageNamespace): Deleted.
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::createWindow const):
* Source/WebCore/page/DOMWindow.cpp:
* Source/WebCore/page/Page.cpp:
(WebCore::m_contentSecurityPolicyModeForExtension):
(WebCore::Page::setSessionID):
(WebCore::Page::sessionStorage): Deleted.
(WebCore::Page::setSessionStorage): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/storage/StorageNamespace.h:
* Source/WebCore/storage/StorageNamespaceProvider.cpp:
(WebCore::StorageNamespaceProvider::sessionStorageArea):
* Source/WebCore/storage/StorageNamespaceProvider.h:
(WebCore::StorageNamespaceProvider::setSessionStorageQuota):
(WebCore::StorageNamespaceProvider::sessionStorageQuota const):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
(WebKit::WebPage::~WebPage):
* Source/WebKit/WebProcess/WebStorage/StorageNamespaceImpl.cpp:
(WebKit::StorageNamespaceImpl::createSessionStorageNamespace):
(WebKit::StorageNamespaceImpl::StorageNamespaceImpl):
(WebKit::StorageNamespaceImpl::copy):
(): Deleted.
(WebKit::StorageNamespaceImpl::~StorageNamespaceImpl): Deleted.
* Source/WebKit/WebProcess/WebStorage/StorageNamespaceImpl.h:
* Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.cpp:
(WebKit::WebStorageNamespaceProvider::incrementUseCount):
(WebKit::WebStorageNamespaceProvider::decrementUseCount):
(WebKit::WebStorageNamespaceProvider::sessionStorageNamespace):
(WebKit::WebStorageNamespaceProvider::copySessionStorageNamespace):
(WebKit::WebStorageNamespaceProvider::createSessionStorageNamespace): Deleted.
* Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.h:
* Source/WebKitLegacy/Storage/StorageNamespaceImpl.h:
* Source/WebKitLegacy/Storage/WebStorageNamespaceProvider.cpp:
(WebKit::WebStorageNamespaceProvider::sessionStorageNamespace):
(WebKit::WebStorageNamespaceProvider::copySessionStorageNamespace):
(WebKit::WebStorageNamespaceProvider::createSessionStorageNamespace): Deleted.
* Source/WebKitLegacy/Storage/WebStorageNamespaceProvider.h:

Canonical link: <a href="https://commits.webkit.org/253762@main">https://commits.webkit.org/253762@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/51341ee40d08ab0f2754512cb0d9da83ca444a71

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86897 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30980 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17769 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95726 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149483 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/90876 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29342 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/25677 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79043 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90950 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92513 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23692 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/73755 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23717 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78687 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/78976 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66722 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27093 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/12826 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27028 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/13840 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2660 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28707 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/36703 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28651 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33119 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/page/Page.cpp:
```diff
@@ -2124,11 +2124,16 @@ void Page::setDebugger(JSC::Debugger* debugger)
+        }
+    } else {
+        if (m_sessionStorage->topLevelOrigin() && !m_sessionStorage->topLevelOrigin()->isSameOriginDomain(topOrigin))
+            m_sessionStorage = m_storageNamespaceProvider->createSessionStorageNamespace(topOrigin, *this, m_settings->sessionStorageQuota());
```
Should we check optionalCreate here too?

sysrqb replied:
Yes, and this wound up simplifying the logic, too. `optionalCreate` is now passed all the way down to `StorageNamespaceImpl` where a new namespace is created - now, it's only created if `optionalCreate` is true.

achristensen07 commented on Source/WebCore/inspector/agents/InspectorDOMStorageAgent.cpp:
```diff
@@ -214,7 +214,7 @@ RefPtr<StorageArea> InspectorDOMStorageAgent::findStorageArea(Protocol::ErrorStr
 
     if (!*isLocalStorage)
-        return m_inspectedPage.sessionStorage()->storageArea(targetFrame->document()->securityOrigin());
+        return m_inspectedPage.sessionStorage(targetFrame->document()->topOrigin())->storageArea(targetFrame->document()->securityOrigin());
```
It may be worth storing auto& document = *targetFrame->document(); for this line and the one below.

sysrqb replied:
Added.

achristensen07 commented on Source/WebCore/page/Chrome.cpp:
```diff
@@ -196,7 +196,7 @@ Page* Chrome::createWindow(Frame& frame, const WindowFeatures& features, const N
 
     if (!features.noopener && !features.noreferrer) {
-        if (auto* oldSessionStorage = m_page.sessionStorage(false))
+        if (auto* oldSessionStorage = m_page.sessionStorage(m_page.mainFrame().document()->topOrigin(), false))
```
It's probably worth null checking the document just to not introduce a new (and probably rare) crash.

sysrqb replied:
Added, and factored out `document`, as well.

achristensen07 commented on Source/WebKit/WebProcess/WebStorage/StorageNamespaceImpl.cpp:
```diff
@@ -42,19 +42,19 @@
     // The identifier of a session storage namespace is the WebPageProxyIdentifier. It is possible we have several WebPage objects in a single process for the same
     // WebPageProxyIdentifier and these need to share the same namespace instance so we know where to route the IPC to.
-    if (auto* existingNamespace = sessionStorageNamespaces().get(identifier))
+    if (auto* existingNamespace = sessionStorageNamespaces().get(std::make_pair(identifier, topLevelOrigin.data())))
```
nit: Can we use initializer lists instead of std::make_pair?  The latter is a bit verbose

sysrqb replied:
Done.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SessionStorage.mm:
```diff
@@ -147,3 +148,187 @@ static void runSessionStorageInNewWindowTest(NSString* createWindowJS, ShouldSes
+
+namespace TestWebKitAPI {
+
+static void synchronouslyLoadUrl(RetainPtr<TestWKWebView>& webView, RetainPtr<TestNavigationDelegate>& delegate, NSString* url)
```
TestWKWebView has synchronouslyLoadRequest

sysrqb replied:
Unfortunately, the test hangs if I use any of the synchronouslyLoad* functions more than once. I didn't immediately see what was causing it, so I created a new wrapper around `loadRequest`. I can dig into this further if you'd prefer that.

achristensen07 replied:
This isn't a blocker for landing this PR.  Look into this later.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SessionStorage.mm:
```diff
@@ -147,3 +148,187 @@ static void runSessionStorageInNewWindowTest(NSString* createWindowJS, ShouldSes
+
+static void verifyTopStorageContent(RetainPtr<TestWKWebView>& webView, RetainPtr<TestNavigationDelegate>& delegate, NSString* topOrigin, NSString* expectedTopContent)
+{
+    // FIXME synchronouslyLoadHTMLString hangs after calling loadRequest
```
I don't think I understand this comment.  We usually don't add commented out code to the repo.  This may be worth investigating separately.

sysrqb replied:
Done. The comment was a bit outdated, anyway.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SessionStorage.mm:
```diff
@@ -147,3 +148,187 @@ static void runSessionStorageInNewWindowTest(NSString* createWindowJS, ShouldSes
+static void verifyPartitionedStorageContent(RetainPtr<TestWKWebView>& webView, RetainPtr<TestNavigationDelegate>& delegate, NSString* topOrigin, NSString* expectedPartitionedContent)
+{
+    // FIXME synchronouslyLoadHTMLString hangs after calling loadRequest
+    // [webView synchronouslyLoadHTMLString:@"Retrieving example.com Session Storage" baseURL:[NSURL URLWithString:seenOiginUrl]];
```
Ditto.  Remove.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SessionStorage.mm:
```diff
@@ -147,3 +148,187 @@ static void runSessionStorageInNewWindowTest(NSString* createWindowJS, ShouldSes
+        { "/frame_storage.html"_s, { frameHTMLGetFrameStorage } },
+    }, HTTPServer::Protocol::HttpsProxy);
+
+    // Inspired by HSTS.mm
```
Unnecessary.

szewai commented on Source/WebCore/page/Page.cpp:
```diff
@@ -2124,11 +2124,11 @@ void Page::setDebugger(JSC::Debugger* debugger)
+    if (!m_sessionStorage || !m_sessionStorage->topLevelOrigin() || !m_sessionStorage->topLevelOrigin()->isSameOriginDomain(topOrigin)) {
         ASSERT(m_settings->sessionStorageQuota() != StorageMap::noQuota);
-        m_sessionStorage = m_storageNamespaceProvider->createSessionStorageNamespace(*this, m_settings->sessionStorageQuota());
+        m_sessionStorage = m_storageNamespaceProvider->maybeCreateSessionStorageNamespace(topOrigin, *this, m_settings->sessionStorageQuota(), optionalCreate);
```
This does not look correct. When a SessionStorageNamespace object is destroyed, its StorageAreaMap objects will be destroyed, and StorageAreaImpl (Storage) objects that weakly points to the StorageAreaMap will be numb. That means when third party iframe accesses sessionStorage, the main frame's existing sessionStorage will be not usable. I think you can write a test to see if it's true.

I think it's better do what TransientLocalStorage does: let StorageNamespaceProvider hold strong ref to SessionStorageNamespace. 
Or if you want SessionStorageNamespace to go away when it's no in use, let StorageNamespaceProvider hold weak ref to SessionStorageNamespace, and let DOMWindow hold strong ref to SessionStorageNamespace.

sysrqb replied:
A third party iframe should have the same namespace as the main frame's namespace because the namespace is keyed on the main frame's origin and the page (session storage) identifier, so I don't believe that should be a problem.

I originally tried replicating how TransientLocalStorage works, but I stopped because I incorrectly thought the Page object and SessionNamespaceProvider object had different lifetimes. I just went back and implemented it, and now I realize this has very similar behavior as letting Page hold the strong ref to the StorageNamespace because Page and StorageNamespaceProvider have the same lifetime (Page holds the strong ref to the namespace provider). This should be the same behavior as the existing implementation in the case where Page is destroyed before DOMWindow, as well. I'm sorry if I'm misunderstanding something.

szewai replied:
Oh you are right that 3rd party frame won't cause SessionStorageNamespace to be destroyed! Sorry for the confusion. But I think cross-origin navigation can still cause the issue. I added a test to the bug https://bugs.webkit.org/show_bug.cgi?id=210776. Please check if it passes with this change.

sysrqb replied:
Thanks for the test! After following the steps I do see the expected results. The item is not accessible on localhost, but it is accessible after "going back" to 127.0.0.1.

sysrqb replied:
You were actually correct about this issue with cross-origin navigation! When I ran the test you wrote, I ran it using a newer version of this changed implementation where the StorageNamespaceProvider holds the strong ref to the SessionStorageNamespace - I didn't test using the branch in this PR. I tested this PR's patch and it does fail your test. I just pushed an updated branch for review and I believe it takes your suggestions and recommendations into account.

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/SessionStorage.mm:
```diff
@@ -147,3 +148,179 @@ static void runSessionStorageInNewWindowTest(NSString* createWindowJS, ShouldSes
+
+namespace TestWebKitAPI {
+
+static void synchronouslyLoadUrl(RetainPtr<TestWKWebView>& webView, RetainPtr<TestNavigationDelegate>& delegate, NSString* url)
```
Is there a reason an API test is preferred than a layout test for this change?
(there's cross-origin test like: LayoutTests/http/tests/security/cross-origin-indexeddb.html)

sysrqb replied:
No, mostly due to still learning, but my understanding is that an API test allows for testing across many origins, whereas layout tests are limited to localhost and loopback address. The API test currently relies on four domains - I can try adding a layout test for this, as well, if you prefer that.

szewai replied:
I think we usually use API test for an API change or something that is untestable with layout tests (e.g. process crash and relaunch). More importantly, tests in SessionStorage.mm only runs on Cocoa platform, and this is not a platform specific change. (I am not sure why we need four domains for testing; it seems {a, b, a/b, b/a} is enough).

sysrqb replied:
That makes sense. This change doesn't need four domains, but I wanted to be sure I didn't miss an edge case. I did find a bug in an earlier version of this patch because I tested both http and https for the same domain. I'll add a layout test.

sysrqb replied:
With the addition of the layout test, should I delete the API test? I believe the API test provides slightly better coverage with both cross-site and cross-origin tests, but the new layout test should be sufficient for the general case.

szewai replied:
I think it's fine to drop the API test

sysrqb replied:
Done.

szewai commented on Source/WebCore/loader/EmptyClients.cpp:
```diff
@@ -505,7 +507,7 @@ class EmptyStorageNamespaceProvider final : public StorageNamespaceProvider {
     };
 
-    Ref<StorageNamespace> createSessionStorageNamespace(Page&, unsigned) final;
+    RefPtr<StorageNamespace> maybeCreateSessionStorageNamespace(SecurityOrigin&, Page&, unsigned, bool) final;
```
The naming and parameter are confusing. I think either we use two functions createSessionStorageNamespace and existingSessionStorageNamespace, or you change this to sessionStorageNamespace(SecurityOrigin&, Page&, ShouldCreateIfNotExists) where ShouldCreateIfNotExists is a enum class.

sysrqb replied:
Initially I created two separate functions as you suggest, but that seemed more complicated, so I re-wrote the patch and renamed the function. I'll make this change.

szewai replied:
If you think two functions are too complicated you can use enum class like
`enum class ShouldCreateIfNotExists : bool { No, Yes };`
We usually do this for readability nowadays

szewai commented on Source/WebKit/WebProcess/WebStorage/StorageNamespaceImpl.cpp:
```diff
@@ -42,19 +42,20 @@
 }
 
-Ref<StorageNamespaceImpl> StorageNamespaceImpl::createSessionStorageNamespace(Identifier identifier, PageIdentifier pageID, unsigned quotaInBytes)
+RefPtr<StorageNamespaceImpl> StorageNamespaceImpl::maybeCreateSessionStorageNamespace(Identifier identifier, PageIdentifier pageID, WebCore::SecurityOrigin& topLevelOrigin, unsigned quotaInBytes, bool optionalCreate)
```
Nothing keeps SessionStorageNamespace alive for the issue mentioned above; so this function will always create new SessionStorageNamespace I think, if optionalCreate is true.

szewai commented on Source/WebCore/loader/EmptyClients.cpp:
```diff
@@ -503,10 +505,12 @@ class EmptyStorageNamespaceProvider final : public StorageNamespaceProvider {
     };
 
-    Ref<StorageNamespace> createSessionStorageNamespace(Page&, unsigned) final;
+    Ref<StorageNamespace> createSessionStorageNamespace(SecurityOrigin&, Page&, unsigned) final;
```
This seems the same as calling sessionStorageNamespace with ShouldCreateIfNotExists::Yes below; maybe we should just remove this

sysrqb replied:
Done. I included createSessionStorageNamespace so SessionStorage remained consistent with LocalStorage, but we can make this simpler by deleting the unnecessary level of indirection.

szewai commented on Source/WebCore/storage/StorageNamespaceProvider.cpp:
```diff
@@ -60,6 +60,16 @@ Ref<StorageArea> StorageNamespaceProvider::localStorageArea(Document& document)
+    ASSERT(document.page());
+
+    Ref<StorageNamespace> storageNamespace = *sessionStorageNamespace(document.topOrigin(), *document.page());
+    return storageNamespace->storageArea(document.securityOrigin());
```
Nit: can be written in one line

sysrqb replied:
Done.

szewai commented on Source/WebKitLegacy/Storage/WebStorageNamespaceProvider.cpp:
```diff
@@ -114,4 +113,10 @@ Ref<StorageNamespace> WebStorageNamespaceProvider::createTransientLocalStorageNa
+    ASSERT(m_sessionStorageQuota != WebCore::StorageMap::noQuota);
+    return createSessionStorageNamespace(topOrigin, page, m_sessionStorageQuota).ptr();
+}
+
```
Why don't we need to hold strong ref to SessionStorageNamespace in WebKitLegacy? It seems nothing holds it alive now (WebCore::Page does not hold strong ref).
We probably should add m_sessionStorageNamespaces to WebStorageNamespaceProvider.

sysrqb replied:
Yes, I was mostly focused on WK2 in that version of the patch. I've added a similar implementation in WebKitLegacy.

szewai commented on Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.h:
```diff
@@ -40,11 +41,16 @@ class WebStorageNamespaceProvider final : public WebCore::StorageNamespaceProvid
+
     const StorageNamespaceIdentifier m_localStorageNamespaceIdentifier;
+
+    HashMap<std::pair<StorageNamespaceImpl::Identifier, WebCore::SecurityOriginData>, RefPtr<WebCore::StorageNamespace>> m_sessionStorageNamespaces;
```
If I understand correctly, WebStorageNamespaceProvider is one per Page, so you don't need StorageNamespaceImpl::Identifier (which is WebPageProxyIdentifier) in the key.

But as mentioned in the comment that:
```
// The identifier of a session storage namespace is the WebPageProxyIdentifier. It is possible we have several WebPage objects in a single process for the same
// WebPageProxyIdentifier and these need to share the same namespace instance so we know where to route the IPC to.
```
We want the map to be shared across StorageNamespaceProviders (as multiple WebPages can have the same WebPageProxyIdentifier). We might make m_sessionStorageNamespaces a static member in WebStorageNamespaceProvider.cpp, with Identifier as the key. WebPage can notify WebStorageNamespaceProvider when it's constructed and destroyed, so WebStorageNamespaceProvider can remove Identifier from map when use count is 0 (No WebPage in the process has the same WebPageProxyIdentifier).

sysrqb replied:
WebPage's WebStorageNamespaceProvider is instantiated in WebPage's constructor as:
`     pageConfiguration.storageNamespaceProvider = WebStorageNamespaceProvider::getOrCreate(*m_pageGroup);`
That resulted in my understanding that the StorageNamespaceProvider is shared across pages within a Page Group (keyed by the PageGroup's pageGroupID). Therefore, I believe the SessionStorageNamespaces map is only isolated by PageGroup, and not per WebPage. I'll make the SessionStorageNamespaces map static in WebStorageNamespaceProvider.cpp, as you suggest, though. I believe we still need both WebPageProxyIdentifier and top-level origin as keys, if my understanding is correct.

szewai replied:
ah, you are right! Actually we have one StorageNamespaceProvider per PageGroup, and I think by default pages are in the same group. So your change is actually fine; you don't need to move m_sessionStorageNamespaces!
But since the WebStorageNamespaceProvider is shared across multiple WebPages, I think we still need a way to destroy StorageNameSpace when WebPage is gone, otherwise there can be abandoned memory (i.e. WebPageProxy is gone but we still keep its SessionStorageNamespace in web process)

What I mean by using Identifier as the key is using Identifier as first-level key to find SessionStorageNamespaces: HashMap<Identifier, SessionStorageNamespaces> and struct SessionStorageNamespaces keeps track of use count (WebPage count, which is updated when WebPage is created and destroyed) and HashMap<SecurityOriginData, StorageNameSpace>. 
This is just an idea of how to get rid of SessionStorageNamespaces that will not be used.

sysrqb replied:
Yes, that's a good idea. My current understanding is that the existing implementation has the same problem, and that is why I didn't worry about it very much (but I'm happy to address it now).

sysrqb replied:
Done.

szewai commented on LayoutTests/http/tests/security/cross-origin-session-storage-third-party-partitioned.html:
```diff
@@ -0,0 +1,105 @@
+            testRunner.waitUntilDone();
+        }
+
+        function log(line) {
```
Not sure if you know this: we have some helper functions in LayoutTests/http/tests/resources/js-test-pre.js, for example you can use debug() for logging, and shouldBeEqualToString() to compare two strings, testPassed()/testFailed for printing something when result is expected/unexpected, etc. I think they will help make the failed case more obvious.

cdumez replied:
I agree it would be nice to write the test with either testharness.js or js-test.js.

Note that js-test-pre.js / js-test-post.js are no longer recommended.

sysrqb replied:
Done.

cdumez commented on Source/WebCore/storage/StorageNamespaceProvider.h:
```diff
@@ -46,15 +46,22 @@ class StorageNamespaceProvider : public RefCounted<StorageNamespaceProvider> {
 protected:
     StorageNamespace* optionalLocalStorageNamespace() { return m_localStorageNamespace.get(); }
 
+    unsigned m_sessionStorageQuota;
```
Data members should be private, not protected. If subclasses need to set /query it, they should be using a getter / setter function.

sysrqb replied:
Done

cdumez commented on Source/WebKitLegacy/Storage/WebStorageNamespaceProvider.cpp:
```diff
@@ -114,4 +108,20 @@ Ref<StorageNamespace> WebStorageNamespaceProvider::createTransientLocalStorageNa
+
+void WebStorageNamespaceProvider::setSessionStorageNamespace(WebCore::SecurityOrigin& topLevelOrigin, WebCore::Page& page, RefPtr<WebCore::StorageNamespace>&& newNamespace)
+{
+    auto& slot = m_sessionStorageNamespaces.add({ &page, topLevelOrigin.data() }, nullptr).iterator->value;
```
What's the purpose of doing this in 2 instructions instead of one, do you want to make sure you overwrite the existing value if the key already existing in the map? If so, you can do so in a single line:
`m_sessionStorageNamespaces.set({ &page, topLevelOrigin.data() }, WTFMove(newNamespace))`

Or if you don't need to overwrite the existing value, just:
`m_sessionStorageNamespaces.add({ &page, topLevelOrigin.data() }, WTFMove(newNamespace))`

sysrqb replied:
Done. This was mostly because I copied a pattern I saw in other places of the codebase, but I agree directly using set() and add() seem better.

cdumez commented on Source/WebKitLegacy/Storage/WebStorageNamespaceProvider.h:
```diff
@@ -49,11 +50,14 @@ class WebStorageNamespaceProvider final : public WebCore::StorageNamespaceProvid
     Ref<WebCore::StorageNamespace> createLocalStorageNamespace(unsigned quota, PAL::SessionID) override;
     Ref<WebCore::StorageNamespace> createTransientLocalStorageNamespace(WebCore::SecurityOrigin&, unsigned quota, PAL::SessionID) override;
 
+    RefPtr<WebCore::StorageNamespace> sessionStorageNamespace(WebCore::SecurityOrigin&, WebCore::Page&, ShouldCreateIfNotExists) override;
```
Could be marked as final

sysrqb replied:
Done - although I'm not sure why this can be final and the others above it aren't, but I'm not opposed to making this final.

cdumez commented on Source/WebKitLegacy/Storage/WebStorageNamespaceProvider.h:
```diff
@@ -49,11 +50,14 @@ class WebStorageNamespaceProvider final : public WebCore::StorageNamespaceProvid
     Ref<WebCore::StorageNamespace> createTransientLocalStorageNamespace(WebCore::SecurityOrigin&, unsigned quota, PAL::SessionID) override;
 
+    RefPtr<WebCore::StorageNamespace> sessionStorageNamespace(WebCore::SecurityOrigin&, WebCore::Page&, ShouldCreateIfNotExists) override;
+    void setSessionStorageNamespace(WebCore::SecurityOrigin&, WebCore::Page&, RefPtr<WebCore::StorageNamespace>&&) override;
```
Could be marked as final

sysrqb replied:
Done.

cdumez commented on Source/WebKitLegacy/Storage/WebStorageNamespaceProvider.h:
```diff
@@ -25,6 +25,7 @@
 
 #pragma once
 
+#include "StorageNamespaceImpl.h"
```
What requires a header include instead of a simple forward declaration? It is not directly obvious to me looking at the new code.

sysrqb replied:
Deleted. That was a stale change - not needed anymore.

cdumez commented on Source/WebKitLegacy/Storage/WebStorageNamespaceProvider.cpp:
```diff
@@ -114,4 +108,20 @@ Ref<StorageNamespace> WebStorageNamespaceProvider::createTransientLocalStorageNa
+
+    auto& slot = m_sessionStorageNamespaces.add({ &page, topLevelOrigin.data() }, nullptr).iterator->value;
+    if (!slot && createIfNotExists == ShouldCreateIfNotExists::Yes)
+        slot = StorageNamespaceImpl::createSessionStorageNamespace(m_sessionStorageQuota, page.sessionID()).ptr();
```
You don't want the `.ptr()` here.

sysrqb replied:
Done.

cdumez commented on Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.h:
```diff
@@ -37,14 +38,26 @@ class WebStorageNamespaceProvider final : public WebCore::StorageNamespaceProvid
     Ref<WebCore::StorageNamespace> createTransientLocalStorageNamespace(WebCore::SecurityOrigin&, unsigned quota, PAL::SessionID) override;
 
+    RefPtr<WebCore::StorageNamespace> sessionStorageNamespace(WebCore::SecurityOrigin&, WebCore::Page&, ShouldCreateIfNotExists) override;
+    void setSessionStorageNamespace(WebCore::SecurityOrigin&, WebCore::Page&, RefPtr<WebCore::StorageNamespace>&&) override;
```
Can be marked as final

sysrqb replied:
Done.

cdumez commented on Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.h:
```diff
@@ -37,14 +38,26 @@ class WebStorageNamespaceProvider final : public WebCore::StorageNamespaceProvid
     Ref<WebCore::StorageNamespace> createLocalStorageNamespace(unsigned quota, PAL::SessionID) override;
     Ref<WebCore::StorageNamespace> createTransientLocalStorageNamespace(WebCore::SecurityOrigin&, unsigned quota, PAL::SessionID) override;
 
+    RefPtr<WebCore::StorageNamespace> sessionStorageNamespace(WebCore::SecurityOrigin&, WebCore::Page&, ShouldCreateIfNotExists) override;
```
Can be marked as final

sysrqb replied:
Done.

cdumez commented on Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.h:
```diff
@@ -37,14 +38,26 @@ class WebStorageNamespaceProvider final : public WebCore::StorageNamespaceProvid
 
+    void incrementUseCount(StorageNamespaceImpl::Identifier);
+    void decrementUseCount(StorageNamespaceImpl::Identifier);
 private:
```
Missing blank line before this.

sysrqb replied:
Done.

cdumez commented on Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.cpp:
```diff
@@ -83,4 +76,43 @@ Ref<WebCore::StorageNamespace> WebStorageNamespaceProvider::createTransientLocal
+void WebStorageNamespaceProvider::decrementUseCount(StorageNamespaceImpl::Identifier identifier)
+{
+    auto& slot = m_sessionStorageNamespaces.add(identifier, SessionStorageNamespaces { }).iterator->value;
+    --slot.useCount;
```
We should add an assertion to make sure the userCount is not already 0 before decrementing.

sysrqb replied:
Done, as well as a couple others.

cdumez commented on Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.cpp:
```diff
@@ -83,4 +76,43 @@ Ref<WebCore::StorageNamespace> WebStorageNamespaceProvider::createTransientLocal
+
+void WebStorageNamespaceProvider::decrementUseCount(StorageNamespaceImpl::Identifier identifier)
+{
+    auto& slot = m_sessionStorageNamespaces.add(identifier, SessionStorageNamespaces { }).iterator->value;
```
Seems weird to call add() to decrement something. We should probably call find() and ASSERT() that the identifier is already in the map.

sysrqb replied:
Done.

cdumez commented on Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.cpp:
```diff
@@ -83,4 +76,43 @@ Ref<WebCore::StorageNamespace> WebStorageNamespaceProvider::createTransientLocal
+
+void WebStorageNamespaceProvider::incrementUseCount(StorageNamespaceImpl::Identifier identifier)
+{
+    auto& slot = m_sessionStorageNamespaces.add(identifier, SessionStorageNamespaces { }).iterator->value;
```
Do we expect the identifier to be already in the map? If so, it would be good to use find() instead of add() and ASSERT() that the identifier is indeed in the map.

sysrqb replied:
Done.

cdumez commented on Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.cpp:
```diff
@@ -83,4 +76,43 @@ Ref<WebCore::StorageNamespace> WebStorageNamespaceProvider::createTransientLocal
+    auto& webPage = WebPage::fromCorePage(page);
+
+    auto& namespacesSlot = m_sessionStorageNamespaces.add(webPage.sessionStorageNamespaceIdentifier(), SessionStorageNamespaces { }).iterator->value;
+    auto& slot = namespacesSlot.map.add(topLevelOrigin.data(), nullptr).iterator->value;
```
`namespacesSlot.map.set(topLevelOrigin.data(), WTFMove(newNamespace));` (if you need to overwrite)
`namespacesSlot.map.add(topLevelOrigin.data(), WTFMove(newNamespace));` (otherwise)

sysrqb replied:
I believe the behavior we want is closer to add(). This function shouldn't be called in a way that replaces an existing StorageNamespace - it should only initialize the namespace for a new window.

cdumez commented on Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.cpp:
```diff
@@ -83,4 +76,43 @@ Ref<WebCore::StorageNamespace> WebStorageNamespaceProvider::createTransientLocal
+    auto& namespacesSlot = m_sessionStorageNamespaces.add(webPage.sessionStorageNamespaceIdentifier(), SessionStorageNamespaces { }).iterator->value;
+    auto& slot = namespacesSlot.map.add(topLevelOrigin.data(), nullptr).iterator->value;
+    if (!slot && createIfNotExists == ShouldCreateIfNotExists::Yes)
+        slot = StorageNamespaceImpl::createSessionStorageNamespace(webPage.sessionStorageNamespaceIdentifier(), webPage.identifier(), topLevelOrigin, m_sessionStorageQuota).ptr();
```
You don't want the `.ptr()` here.

sysrqb replied:
Done.

cdumez commented on Source/WebKit/WebProcess/WebStorage/StorageNamespaceImpl.cpp:
```diff
@@ -42,19 +42,9 @@
 using namespace WebCore;
 
-static HashMap<StorageNamespaceImpl::Identifier, StorageNamespaceImpl*>& sessionStorageNamespaces()
+Ref<StorageNamespaceImpl> StorageNamespaceImpl::createSessionStorageNamespace(Identifier identifier, PageIdentifier pageID, WebCore::SecurityOrigin& topLevelOrigin, unsigned quotaInBytes)
```
Can we pass the topLevelOrigin as a `const WebCore::SecurityOrigin&` ?

sysrqb replied:
Done.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -1092,6 +1094,8 @@ WebPage::~WebPage()
         m_footerBanner->detachFromPage();
 #endif
 
+    WebStorageNamespaceProvider::getOrCreate(*m_pageGroup)->decrementUseCount(sessionStorageNamespaceIdentifier());
```
Seems a little weird that we're calling something that might "create" a provider, just to decrement something.

sysrqb replied:
I'll respond to the other comments tomorrow, but this is the last improvement that I'm still thinking about. I don't really like this, either, but I'm not yet sure which alternative we should use. Originally, I naively tried `static_cast<WebStorageNamespaceProvider&>(m_page->storageNamespaceProvider()).decrementUseCount(sessionStorageNamespaceIdentifier());`, but that doesn't work because `m_page` is destroyed before we enter the WebPage's destructor. We could let WebPage hold its own reference to the StorageNamespaceProvider, but that's also not ideal. We could also create a static `decrementUseCount` method that operates like the current `getOrCreate()->decrement...()` call, but only if the Namespace currently exists. I would prefer symmetric `increment` and `decrement` functions, but incrementing the useCount seems to be much simpler than decrementing.

sysrqb replied:
I went with using the latter. As it turned out, calling getOrCreate() was messy and created a problem in an edge case where it actually did create a new Provider and then tried to decrement the counter.

cdumez commented on Source/WebCore/storage/StorageNamespaceProvider.h:
```diff
@@ -46,15 +46,22 @@ class StorageNamespaceProvider : public RefCounted<StorageNamespaceProvider> {
     WEBCORE_EXPORT void setSessionIDForTesting(PAL::SessionID);
 
+    void setSessionStorageQuota(unsigned quota) { m_sessionStorageQuota = quota; }
+    virtual void setSessionStorageNamespace(SecurityOrigin&, Page&, RefPtr<StorageNamespace>&&) = 0;
```
Can we pass the origin as a `const SecurityOrigin&` ?

sysrqb replied:
Done.

cdumez commented on Source/WebCore/storage/StorageNamespaceProvider.h:
```diff
@@ -46,15 +46,22 @@ class StorageNamespaceProvider : public RefCounted<StorageNamespaceProvider> {
+    Ref<StorageArea> sessionStorageArea(Document&);
+
+    enum class ShouldCreateIfNotExists : bool { No, Yes };
+    virtual RefPtr<StorageNamespace> sessionStorageNamespace(SecurityOrigin&, Page&, ShouldCreateIfNotExists = ShouldCreateIfNotExists::Yes) = 0;
```
Can we pass the origin as a const SecurityOrigin& ?

sysrqb replied:
Done.

cdumez commented on Source/WebCore/storage/StorageNamespaceProvider.h:
```diff
@@ -46,15 +46,22 @@ class StorageNamespaceProvider : public RefCounted<StorageNamespaceProvider> {
     Ref<StorageArea> localStorageArea(Document&);
+    Ref<StorageArea> sessionStorageArea(Document&);
+
+    enum class ShouldCreateIfNotExists : bool { No, Yes };
```
`ShouldCreateIfNotExists` sounds a little weird. I guess `ShouldCreateIfItDoesNotExist` would be less weird but long. Maybe `ShouldCreateIfNecessary`?

sysrqb replied:
Renamed to `ShouldCreateNamespace` - following the lead of `NetworkStorageManager::ShouldWriteOriginFile`.

sysrqb replied:
Now ShouldCreateNamespace is only needed by WebKitLegacy, as well.

cdumez commented on Source/WebCore/loader/EmptyClients.cpp:
```diff
@@ -494,6 +494,8 @@ class EmptyStorageNamespaceProvider final : public StorageNamespaceProvider {
         }
+
+        SecurityOrigin* topLevelOrigin() const final { return nullptr; };
     private:
```
Missing blank line before this.

sysrqb replied:
Done.

szewai commented on Source/WebKitLegacy/Storage/WebStorageNamespaceProvider.h:
```diff
@@ -49,11 +49,14 @@ class WebStorageNamespaceProvider final : public WebCore::StorageNamespaceProvid
+    void setSessionStorageNamespace(const WebCore::SecurityOrigin&, WebCore::Page&, RefPtr<WebCore::StorageNamespace>&&) final;
+
     const String m_localStorageDatabasePath;
+    HashMap<std::pair<WebCore::Page *, WebCore::SecurityOriginData>, RefPtr<WebCore::StorageNamespace>> m_sessionStorageNamespaces;
```
I think we want to avoid using raw pointer in key. Can we change it to something like WeakHashMap<WebCore::Page, HashMap<WebCore::SecurityOriginData, RefPtr<WebCore::StorageNamespace>>>?

sysrqb replied:
Thanks, I used that. I saw other uses of raw pointers in map keys, but avoiding new usage seems like a good idea.

szewai commented on Source/WebCore/page/Chrome.cpp:
```diff
@@ -195,9 +196,11 @@ Page* Chrome::createWindow(Frame& frame, const WindowFeatures& features, const N
+    if (!features.noopener && !features.noreferrer && document) {
+        constexpr auto doNotCreate = StorageNamespaceProvider::ShouldCreateNamespace::No;
+        if (auto oldSessionStorage = m_page.storageNamespaceProvider().sessionStorageNamespace(document->topOrigin(), m_page, doNotCreate))
+            newPage->storageNamespaceProvider().setSessionStorageNamespace(document->topOrigin(), *newPage, oldSessionStorage->copy(*newPage));
```
I think we need to copy all SessionStorageNamespaces from current page to new page, according to [spec](https://html.spec.whatwg.org/multipage/webstorage.html#dom-sessionstorage):
While [creating a new auxiliary browsing context](https://html.spec.whatwg.org/multipage/browsers.html#creating-a-new-auxiliary-browsing-context), the session storage [is copied](https://html.spec.whatwg.org/multipage/browsers.html#copy-session-storage) over.

That means we need to copy all SessionStorage items from one page to another, e.g.
probably by changing setSessionStorageNamespace to setSessionStorageNamespaces (or name it copySessionStorageNamespaces), which takes two Pages as parameters, and copies each SessionStorageNamespace of the first page to seconds Page's SessionStorageNamespaces.

szewai replied:
One thing to notice is that you might need to move 
```
if (auto networkProcessConnection = WebProcess::singleton().existingNetworkProcessConnection())
        networkProcessConnection->connection().send(Messages::NetworkStorageManager::CloneSessionStorageNamespace(m_storageNamespaceID, WebPage::fromCorePage(newPage).sessionStorageNamespaceIdentifier()), 0);
```
from StorageNamespaceImpl::copy to the WebStorageNamespaceProvider::copySessionStorageNamespaces function, because this message just tells network process to copy the page's StorageNamespaces to new page -- we can send one message per page copy, not per StorageNamespace copy.

sysrqb replied:
Thanks for noticing I missed this! These are good suggestions, I followed them.

szewai reviewed (commented):
The patch overall looks reasonable to me; just two small changes

szewai commented on Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.cpp:
```diff
@@ -83,4 +103,58 @@ Ref<WebCore::StorageNamespace> WebStorageNamespaceProvider::createTransientLocal
+    if (auto networkProcessConnection = WebProcess::singleton().existingNetworkProcessConnection())
+        networkProcessConnection->connection().send(Messages::NetworkStorageManager::CloneSessionStorageNamespace(srcWebPage.sessionStorageNamespaceIdentifier(), dstWebPage.sessionStorageNamespaceIdentifier()), 0);
+
+    for (auto& [k, v] : srcNamespacesMap)
```
Nit: maybe `[origin, namespace]` instead of `[k, v]`.
(I think we prefer using full name for clarity as in https://webkit.org/code-style-guidelines/.)

sysrqb replied:
Done

szewai commented on Source/WebKitLegacy/Storage/WebStorageNamespaceProvider.cpp:
```diff
@@ -114,4 +109,45 @@ Ref<StorageNamespace> WebStorageNamespaceProvider::createTransientLocalStorageNa
+
+    HashMap<SecurityOriginData, RefPtr<StorageNamespace>> map;
+    auto& dstPageSessionStorageNamespaces = dstSessionStorageNamespaces.add(dstPage, map).iterator->value;
+    for (auto& [k, v] : srcPageSessionStorageNamespaces)
```
Ditto.

szewai reviewed (approved):
LGTM

achristensen07 commented on Source/WebCore/storage/StorageNamespaceProvider.h:
```diff
@@ -65,6 +71,8 @@ class StorageNamespaceProvider : public RefCounted<StorageNamespaceProvider> {
     RefPtr<StorageNamespace> m_localStorageNamespace;
     HashMap<SecurityOriginData, RefPtr<StorageNamespace>> m_transientLocalStorageNamespaces;
+
+    unsigned m_sessionStorageQuota;
```
Can we initialize this to 0 instead of using uninitialized memory if setSessionStorageQuota isn't called?

sysrqb replied:
Yes, good catch.

achristensen07 commented on Source/WebKit/WebProcess/WebStorage/StorageNamespaceImpl.cpp:
```diff
@@ -67,27 +57,14 @@ Ref<StorageNamespaceImpl> StorageNamespaceImpl::createTransientLocalStorageNames
     , m_storageNamespaceID(storageNamespaceID)
     , m_sessionPageID(pageIdentifier)
-    , m_topLevelOrigin(topLevelOrigin)
+    , m_topLevelOrigin(const_cast<WebCore::SecurityOrigin *>(topLevelOrigin))
```
Can we make m_topLevelOrigin a. RefPtr<const WebCore::SecurityOrigin> instead of having a const_cast?

sysrqb replied:
Done, made it `const`.

sysrqb committed:
Partition sessionStorage by top-level origin
https://bugs.webkit.org/show_bug.cgi?id=210776
rdar://57674840

Reviewed by Alex Christensen and Sihui Liu.

A page could access its origin's sessionStorage in both first-party and
third-party contexts. This differs from other storage systems and caches where
in a third-party context those storage areas are partitioned based on the
first-party (top-level) origin. This issue is addressed by ensuring that a
Window's sessionStorage takes the top-level origin into account at access-time.

This accounting is accomplished by keying the StorageNamespace by top-level
origin and the namespace identifier. One complexity was how the Page object
caches the resulting StorageNamespace, now partitioned by origin. This change
interacted poorly with the creation and destruction of the Page object during
navigation. As a result, ownership of the StorageNamespace is moved from the
Page into the StorageNamespaceProvider.

* LayoutTests/http/tests/security/cross-origin-session-storage-third-party-partitioned-expected.txt: Added.
* LayoutTests/http/tests/security/cross-origin-session-storage-third-party-partitioned.html: Added.
* LayoutTests/http/tests/security/resources/cross-origin-iframe-for-partitioned-session-storage.html: Added.
* LayoutTests/http/tests/security/resources/cross-origin-main-frame-for-partitioned-session-storage.html: Added.
* Source/WebCore/inspector/agents/InspectorDOMStorageAgent.cpp:
(WebCore::InspectorDOMStorageAgent::findStorageArea):
* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::EmptyStorageNamespaceProvider::sessionStorageNamespace):
(WebCore::EmptyStorageNamespaceProvider::createSessionStorageNamespace): Deleted.
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::createWindow const):
* Source/WebCore/page/DOMWindow.cpp:
* Source/WebCore/page/Page.cpp:
(WebCore::m_contentSecurityPolicyModeForExtension):
(WebCore::Page::setSessionID):
(WebCore::Page::sessionStorage): Deleted.
(WebCore::Page::setSessionStorage): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/storage/StorageNamespace.h:
* Source/WebCore/storage/StorageNamespaceProvider.cpp:
(WebCore::StorageNamespaceProvider::sessionStorageArea):
* Source/WebCore/storage/StorageNamespaceProvider.h:
(WebCore::StorageNamespaceProvider::setSessionStorageQuota):
(WebCore::StorageNamespaceProvider::sessionStorageQuota const):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
(WebKit::WebPage::~WebPage):
* Source/WebKit/WebProcess/WebStorage/StorageNamespaceImpl.cpp:
(WebKit::StorageNamespaceImpl::createSessionStorageNamespace):
(WebKit::StorageNamespaceImpl::StorageNamespaceImpl):
(WebKit::StorageNamespaceImpl::copy):
(): Deleted.
(WebKit::StorageNamespaceImpl::~StorageNamespaceImpl): Deleted.
* Source/WebKit/WebProcess/WebStorage/StorageNamespaceImpl.h:
* Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.cpp:
(WebKit::WebStorageNamespaceProvider::incrementUseCount):
(WebKit::WebStorageNamespaceProvider::decrementUseCount):
(WebKit::WebStorageNamespaceProvider::sessionStorageNamespace):
(WebKit::WebStorageNamespaceProvider::copySessionStorageNamespace):
(WebKit::WebStorageNamespaceProvider::createSessionStorageNamespace): Deleted.
* Source/WebKit/WebProcess/WebStorage/WebStorageNamespaceProvider.h:
* Source/WebKitLegacy/Storage/StorageNamespaceImpl.h:
* Source/WebKitLegacy/Storage/WebStorageNamespaceProvider.cpp:
(WebKit::WebStorageNamespaceProvider::sessionStorageNamespace):
(WebKit::WebStorageNamespaceProvider::copySessionStorageNamespace):
(WebKit::WebStorageNamespaceProvider::createSessionStorageNamespace): Deleted.
* Source/WebKitLegacy/Storage/WebStorageNamespaceProvider.h:

Canonical link: https://commits.webkit.org/253762@main
